### PR TITLE
support toolimage when backup/restore using dumpling/lightning

### LINKED ISSUE #3619: Support setting tool image for backup with dumpling and restore with lightning
## Feature Request

**Is your feature request related to a problem? Please describe:**

**Describe the feature you'd like:**

https://github.com/pingcap/tidb-operator/pull/3471 has supported setting BR image during backup and restore with BR.
We can leverage the `toolImage` field added to support setting dumping image during backup and lightning image during restore.
The default image tag can be the tidb image tag.
**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by BinChenn**
### What problem does this PR solve?
Support setting `spec.toolImage` for backup with dumpling and restore with lightning.

Close #3619

### What is changed and how does it work?
If `spec.toolImage` is configured when using dumpling/lightning, add initContainers to copy binary file `dumpling`/`tidb-lightning` from toolImage and run this binary file to backup/restore.

### Code changes

- [x] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [x] Unit test 

- [ ] E2E test 

- [x] Manual test 

    1. Backup using dumpling when configuring `spec.toolImage`

        Prepare tidb cluster. 

        ```shell
        kind create cluster
        ./hack/local-up-operator.sh
        kubectl apply -f example/basic/tidb-cluster
        ```

        Backup to gcs suing dumpling according to [docs](https://docs.pingcap.com/zh/tidb-in-kubernetes/stable/backup-to-gcs), configuration is as follows.

        ```yaml
        ---
        apiVersion: pingcap.com/v1alpha1
        kind: Backup
        metadata:
          name: demo1-backup-gcs
          namespace: default
        spec:
          toolImage: pingcap/dumpling:v4.0.8
          from:
            host: basic-tidb.default
            port: 4000
            user: root
            secretName: backup-demo1-tidb-secret
          gcs:
            secretName: gcs-secret
            projectId: ${project_id}
            bucket: ${bucket}
            prefix: ${prefix}
          storageClassName: standard
          storageSize: 1Gi
        ```

        Backup pod logs show below, we can see that dumpling command is `/var/lib/dumpling-bin/dumpling` backup successfully.

        ```
        I1223 07:27:34.672348       1 export.go:86] The dump process is ready, command "/var/lib/dumpling-bin/dumpling --output=${path} --host=basic-tidb.default --port=4000 --user=root --password= --filter *.* --filter !/^(mysql|test|INFORMATION_SCHEMA|PERFORMANCE_SCHEMA|METRICS_SCHEMA|INSPECTION_SCHEMA)$/.* --threads=16 --rows=10000"
        I1223 07:27:40.657449       1 manager.go:348] backup cluster default/demo1-backup-gcs data to gcs success
        I1223 07:27:40.683256       1 backup_status_updater.go:64] Backup: [default/demo1-backup-gcs] updated successfully
        ```

    2. Restore using tidb-lightning when configuring `spec.toolImage`

        ```yaml
        ---
        apiVersion: pingcap.com/v1alpha1
        kind: Restore
        metadata:
          name: demo2-restore
          namespace: default
        spec:
          toolImage: pingcap/tidb-lightning:v4.0.8
          to:
            host: basic-tidb.default
            port: 4000
            user: root
            secretName: backup-demo1-tidb-secret
          gcs:
            projectId: ${project_id}
            secretName: gcs-secret
            path: gcs://${path}
          storageSize: 1Gi
        ```

        Restore pod logs show below, we can see that command is `/var/lib/lightning-bin/tidb-lightning`, restore successfully.

        ```
        I1223 07:39:21.850310       1 import.go:119] The lightning process is ready, command "/var/lib/lightning-bin/tidb-lightning --status-addr=0.0.0.0:8289 --backend=tidb --server-mode=false --log-file=- --tidb-user=root --tidb-password= --tidb-host=basic-tidb.default --tidb-port=4000"
        I1223 07:39:22.065700       1 manager.go:163] restore cluster default/demo2-restore from backup gcs://${path} success
        I1223 07:39:22.094951       1 restore_status_updater.go:64] Restore: [default/demo2-restore] updated successfully
        ```

    3. Backup/Restore using dumpling/tidb-lightning without configuring `spec.toolImage`

        If we do not set `spec.toolImage` when using dumpling/tidb-lightning, backup-manager will execute binary file `dumpling`/`tidb-lightning` in `tidb-backup-manager` image. Pod logs will show as below, we can see that command is `/dumpling` and `/tidb-lightning` .

        ```
        I1223 07:22:24.858943       1 export.go:86] The dump process is ready, command "/dumpling --output=${path} --host=basic-tidb.default --port=4000 --user=root --password= --filter *.* --filter !/^(mysql|test|INFORMATION_SCHEMA|PERFORMANCE_SCHEMA|METRICS_SCHEMA|INSPECTION_SCHEMA)$/.* --threads=16 --rows=10000"
        I1223 07:22:30.581400       1 manager.go:348] backup cluster default/demo1-backup-gcs data to gcs success
        I1223 07:22:30.610813       1 backup_status_updater.go:64] Backup: [default/demo1-backup-gcs] updated successfully
        ```

        ```
        I1223 07:35:17.027269       1 import.go:119] The lightning process is ready, command "/tidb-lightning --status-addr=0.0.0.0:8289 --backend=tidb --server-mode=false --log-file=- --tidb-user=root --tidb-password= --tidb-host=basic-tidb.default --d=${path} --tidb-port=4000"
        I1223 07:35:17.747189       1 manager.go:163] restore cluster default/demo2-restore from backup gcs://${path} success
        I1223 07:35:17.774306       1 restore_status_updater.go:64] Restore: [default/demo2-restore] updated successfully
        ```

        

- [ ] No code 

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [ ] Need to cherry-pick to the release branch 
- [ ] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note
Support `spec.toolImage` for `Backup` & `Restore` with Dumpling/TiDB Lightning to define the image used to provide the Dumpling/TiDB Lightning binary executables.
```

### CODE REVIEW

File: cmd/backup-manager/app/export/export.go [hunk 1]
```diff
@@ -78,9 +78,14 @@ func (bo *Options) dumpTidbClusterData(backup *v1alpha1.Backup) (string, error)
 		args = append(args, fmt.Sprintf("--key=%s", path.Join(util.TiDBClientTLSPath, corev1.TLSPrivateKeyKey)))
 	}
 
-	klog.Infof("The dump process is ready, command \"/dumpling %s\"", strings.Join(args, " "))
+	bin := "/dumpling"
+	if backup.Spec.ToolImage != "" {
+		bin = path.Join(util.DumplingBinPath, "dumpling")
+	}
+
+	klog.Infof("The dump process is ready, command \"%s %s\"", bin, strings.Join(args, " "))
```
Comment (dragonly) [inline cmd/backup-manager/app/export/export.go:86]:
```suggestion
	binPath := "/dumpling"
	if backup.Spec.ToolImage != "" {
		binPath = path.Join(util.DumplingBinPath, "dumpling")
	}

	klog.Infof("The dump process is ready, command \"%s %s\"", binPath, strings.Join(args, " "))
```

File: cmd/backup-manager/app/export/export.go [hunk 2]
```diff
@@ -78,9 +78,14 @@ func (bo *Options) dumpTidbClusterData(backup *v1alpha1.Backup) (string, error)
...
 
-	output, err := exec.Command("/dumpling", args...).CombinedOutput()
+	output, err := exec.Command(bin, args...).CombinedOutput()
```
Comment (dragonly) [inline cmd/backup-manager/app/export/export.go:88]:
```suggestion
	output, err := exec.Command(binPath, args...).CombinedOutput()
```

File: cmd/backup-manager/app/import/import.go [hunk 3]
```diff
@@ -108,7 +111,14 @@ func (ro *Options) loadTidbClusterData(restorePath string, tableFilter []string)
...
 
-	output, err := exec.Command("/tidb-lightning", args...).CombinedOutput()
+	bin := "/tidb-lightning"
+	if restore.Spec.ToolImage != "" {
+		bin = path.Join(util.LightningBinPath, "tidb-lightning")
+	}
+
+	klog.Infof("The lightning process is ready, command \"%s %s\"", bin, strings.Join(args, " "))
+
+	output, err := exec.Command(bin, args...).CombinedOutput()
```
Comment (dragonly) [inline cmd/backup-manager/app/import/import.go:121]:
```suggestion
	binPath := "/tidb-lightning"
	if restore.Spec.ToolImage != "" {
		binPath = path.Join(util.LightningBinPath, "tidb-lightning")
	}

	klog.Infof("The lightning process is ready, command \"%s %s\"", binPath, strings.Join(args, " "))

	output, err := exec.Command(binPath, args...).CombinedOutput()
```

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 4]
```diff
@@ -1278,7 +1278,7 @@ type BackupSpec struct {
 	// Base tolerations of backup Pods, components may add more tolerations upon this respectively
 	// +optional
 	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
-	// ToolImage specifies the tool image used in the backup/restore, only BR image is supported for now
+	// ToolImage specifies the tool image used in the backup/restore, supporting BR image and Lightning/Dumpling images
```
Comment (dragonly) [inline pkg/apis/pingcap/v1alpha1/types.go:1281]:
```suggestion
	// ToolImage specifies the tool image used in `Backup` and `Restore`, which supports BR, Lightning and Dumpling images
```

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 5]
```diff
@@ -1539,7 +1539,7 @@ type RestoreSpec struct {
 	UseKMS bool `json:"useKMS,omitempty"`
 	// Specify service account of restore
 	ServiceAccount string `json:"serviceAccount,omitempty"`
-	// ToolImage specifies the tool image used in the backup/restore, only BR image is supported for now
+	// ToolImage specifies the tool image used in the backup/restore, supporting BR image and Lightning/Dumpling images
```
Comment (dragonly) [inline pkg/apis/pingcap/v1alpha1/types.go:1542]:
```suggestion
	// ToolImage specifies the tool image used in `Backup` and `Restore`, which supports BR, Lightning and Dumpling images
```

File: cmd/backup-manager/app/export/export.go (hunk 1 above)
Comment (BinChenn) [reply cmd/backup-manager/app/export/export.go]: fixed

File: cmd/backup-manager/app/export/export.go (hunk 2 above)
Comment (BinChenn) [reply cmd/backup-manager/app/export/export.go]: fixed

File: cmd/backup-manager/app/import/import.go (hunk 3 above)
Comment (BinChenn) [reply cmd/backup-manager/app/import/import.go]: fixed

File: pkg/apis/pingcap/v1alpha1/types.go (hunk 4 above)
Comment (BinChenn) [reply pkg/apis/pingcap/v1alpha1/types.go]: fixed

File: pkg/apis/pingcap/v1alpha1/types.go (hunk 5 above)
Comment (BinChenn) [reply pkg/apis/pingcap/v1alpha1/types.go]: fixed

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 6]
```diff
@@ -1278,7 +1278,7 @@ type BackupSpec struct {
 	// Base tolerations of backup Pods, components may add more tolerations upon this respectively
 	// +optional
 	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
-	// ToolImage specifies the tool image used in the backup/restore, only BR image is supported for now
+	// ToolImage specifies the tool image used in `Backup` and `Restore`, which supports BR, Lightning and Dumpling images
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/types.go:1281]: Please give examples for the two tools.

Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/types.go:1281]:
```suggestion
	// ToolImage specifies the tool image used in `Backup` and `Restore`, which supports BR and Dumpling images
```

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 7]
```diff
@@ -1539,7 +1539,7 @@ type RestoreSpec struct {
 	UseKMS bool `json:"useKMS,omitempty"`
 	// Specify service account of restore
 	ServiceAccount string `json:"serviceAccount,omitempty"`
-	// ToolImage specifies the tool image used in the backup/restore, only BR image is supported for now
+	// ToolImage specifies the tool image used in `Backup` and `Restore`, which supports BR, Lightning and Dumpling images
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/types.go:1542]:
```suggestion
	// ToolImage specifies the tool image used in `Backup` and `Restore`, which supports BR and Lightning images
```

Comment (DanielZhangQD) [reply pkg/apis/pingcap/v1alpha1/types.go:1542]: Also add examples here.

File: cmd/backup-manager/app/import/manager.go [hunk 8]
```diff
@@ -147,7 +147,7 @@ func (rm *RestoreManager) performRestore(restore *v1alpha1.Restore) error {
 	}
 	klog.Infof("get cluster %s commitTs %s success", rm, commitTs)
 
-	err = rm.loadTidbClusterData(unarchiveDataPath, restore.Spec.TableFilter)
+	err = rm.loadTidbClusterData(unarchiveDataPath, restore)
```
[the remark this answers is not in the document]
Comment (BinChenn) [reply cmd/backup-manager/app/import/manager.go:150]: We need to access `restore.Spec.toolImage` in func `loadTidbClusterData`.

File: pkg/apis/pingcap/v1alpha1/types.go (hunk 6 above)
Comment (BinChenn) [reply pkg/apis/pingcap/v1alpha1/types.go]: fixed

Comment (BinChenn) [reply pkg/apis/pingcap/v1alpha1/types.go]: fixed

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 9]
```diff
@@ -1278,7 +1278,7 @@ type BackupSpec struct {
 	// Base tolerations of backup Pods, components may add more tolerations upon this respectively
 	// +optional
 	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
-	// ToolImage specifies the tool image used in the backup/restore, only BR image is supported for now
+	// ToolImage specifies the tool image used in `Backup` and `Restore`, which supports BR, and Dumpling images. For examples `spec.toolImage: pingcap/br:v4.0.8` or `spec.toolImage: pingcap/dumpling:v4.0.8`
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/types.go:1281]:
```suggestion
	// ToolImage specifies the tool image used in `Backup`, which supports BR and Dumpling images.
	// For examples `spec.toolImage: pingcap/br:v4.0.8` or `spec.toolImage: pingcap/dumpling:v4.0.8`
```

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 10]
```diff
@@ -1539,7 +1539,7 @@ type RestoreSpec struct {
 	UseKMS bool `json:"useKMS,omitempty"`
 	// Specify service account of restore
 	ServiceAccount string `json:"serviceAccount,omitempty"`
-	// ToolImage specifies the tool image used in the backup/restore, only BR image is supported for now
+	// ToolImage specifies the tool image used in `Backup` and `Restore`, which supports BR and TiDB Lightning images. For examples `spec.toolImage: pingcap/br:v4.0.8` or `spec.toolImage: pingcap/tidb-lightning:v4.0.8`
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/types.go:1542]:
```suggestion
	// ToolImage specifies the tool image used in `Restore`, which supports BR and TiDB Lightning images.
	// For examples `spec.toolImage: pingcap/br:v4.0.8` or `spec.toolImage: pingcap/tidb-lightning:v4.0.8`
```

File: pkg/apis/pingcap/v1alpha1/types.go (hunk 9 above)
Comment (BinChenn) [reply pkg/apis/pingcap/v1alpha1/types.go]: fixed

File: pkg/apis/pingcap/v1alpha1/types.go (hunk 10 above)
Comment (BinChenn) [reply pkg/apis/pingcap/v1alpha1/types.go]: fixed

**Review by csuzhangxc — APPROVED** [review]
LGTM

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 11]
```diff
@@ -1278,7 +1278,8 @@ type BackupSpec struct {
 	// Base tolerations of backup Pods, components may add more tolerations upon this respectively
 	// +optional
 	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
-	// ToolImage specifies the tool image used in `Backup` and `Restore`, which supports BR, and Dumpling images. For examples `spec.toolImage: pingcap/br:v4.0.8` or `spec.toolImage: pingcap/dumpling:v4.0.8`
+	// ToolImage specifies the tool image used in `Backup` and `Restore`, which supports BR, and Dumpling images.
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/types.go:1281]:
```suggestion
	// ToolImage specifies the tool image used in `Backup`, which supports BR and Dumpling images.
```

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 12]
```diff
@@ -1539,7 +1540,8 @@ type RestoreSpec struct {
 	UseKMS bool `json:"useKMS,omitempty"`
 	// Specify service account of restore
 	ServiceAccount string `json:"serviceAccount,omitempty"`
-	// ToolImage specifies the tool image used in `Backup` and `Restore`, which supports BR and TiDB Lightning images. For examples `spec.toolImage: pingcap/br:v4.0.8` or `spec.toolImage: pingcap/tidb-lightning:v4.0.8`
+	// ToolImage specifies the tool image used in `Backup` and `Restore`, which supports BR and TiDB Lightning images.
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/types.go:1543]:
```suggestion
	// ToolImage specifies the tool image used in `Restore`, which supports BR and TiDB Lightning images.
```

**Review by DanielZhangQD — APPROVED** [review]
LGTM

**Pull request merged**